[exporter/loki] Refactor the Loki exporter

jpkrohling opened the pull request:
This commit brings a new implementation of the Loki exporter while
maintaining compatibility with the previous version (called "legacy" in
this PR). The new exporter has a cleaner logic, allowing the Loki labels
to be configured as part of the data points instead of being
part of the static configuration. This will be useful in the future when
migrating from the Loki exporter to a native OTLP endpoint in Loki.

Fixes #12873

Signed-off-by: Juraci Paixão Kröhling <juraci@kroehling.de>

jpkrohling committed:
[exporter/loki] Refactor the Loki exporter

This commit brings a new implementation of the Loki exporter while
maintaining compatibility with the previous version (called "legacy" in
this PR). The new exporter has a cleaner logic, allowing the Loki labels
to be configured as part of the data points themselves instead of being
part of the static configuration. This will be useful in the future when
migrating from the Loki exporter to a native OTLP endpoint in Loki.

Fixes #12873

Signed-off-by: Juraci Paixão Kröhling <juraci@kroehling.de>

jpkrohling commented:
@gramidt, as the original author of this exporter, would you be able to take a look at this? 

@kovrus, would you also be able to take a look at this one?

kovrus commented on exporter/lokiexporter/convert.go:
```diff
@@ -0,0 +1,107 @@
+	return out
+}
+
+func convertAttributesToLabels(logger *zap.Logger, attributes pcommon.Map, attrsToSelect pcommon.Value) model.LabelSet {
```
`logger` is passed but not used.

kovrus commented on exporter/lokiexporter/config_test.go:
```diff
@@ -15,515 +15,84 @@
-	"go.opentelemetry.io/collector/service/servicetest"
 )
 
-func TestLoadConfig(t *testing.T) {
```
Let's maybe keep it the `TestLoadConfig` for the new exporter. We can also use `confmaptest` 
 https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/13224 to read the config file.

kovrus commented on exporter/lokiexporter/next_exporter.go:
```diff
@@ -0,0 +1,177 @@
+}
+
+func (l *nextLokiExporter) pushLogData(ctx context.Context, ld plog.Logs) error {
+	pushReq, _ := logDataToLoki(l.settings.Logger, ld)
```
`numDroppedLogs` is skipped here, maybe we can log  it here or remove from return args of `logDataToLoki`.

jpkrohling replied:
I would remove it; we can reinstate that once we can record metrics on the components.

kovrus reviewed (commented):
lgtm 👍 

left a couple of minor comments

jpkrohling committed:
First review changes

- Reinstated the test load config for the new exporter
- Removed logger parameter where not necessary
- Removed num of dropped logs, should be a metric in the future

Signed-off-by: Juraci Paixão Kröhling <juraci@kroehling.de>

jpkrohling commented:
If this looks OK, I'll fix the merge conflicts.

jpkrohling committed:
Add changelog entry

Signed-off-by: Juraci Paixão Kröhling <juraci@kroehling.de>

jpkrohling committed:
lint

Signed-off-by: Juraci Paixão Kröhling <juraci@kroehling.de>

jpkrohling committed:
make impi

Signed-off-by: Juraci Paixão Kröhling <juraci@kroehling.de>

bogdandrutu commented on exporter/lokiexporter/factory_next.go:
```diff
@@ -0,0 +1,39 @@
+	"go.opentelemetry.io/collector/exporter/exporterhelper"
+)
+
+func createNextLogsExporter(ctx context.Context, set component.ExporterCreateSettings, cfg *Config) (component.LogsExporter, error) {
```
I would keep the name "createLogsExporter" and move it to factory.go.

jpkrohling replied:
The default route is still to use the "legacy", which is the reason I chose "next" as the name for this. I'll eventually change the code path to use this by default and rename it to reflect the change.

bogdandrutu commented on exporter/lokiexporter/next_exporter.go:
```diff
@@ -0,0 +1,176 @@
+
+				mergedLabels := convertAttributesAndMerge(log.Attributes(), resource.Attributes())
+				// remove the attributes that were promoted to labels
+				removeAttributes(log.Attributes(), mergedLabels)
```
Looked at the implementation of this, it is faster to use `RemoveIf` instead of multiple remove calls.

jpkrohling commented on exporter/lokiexporter/config_deprecated.go:
```diff
@@ -0,0 +1,98 @@
+	"github.com/prometheus/common/model"
+)
+
+type Tenant struct {
```
Good point. I thought it wouldn't be needed, as the place they are used within the config is marked as deprecated, but you are right. People might be using this outside of that context, they deserve to see the deprecation warning as well.

jpkrohling commented on exporter/lokiexporter/config_deprecated.go:
```diff
@@ -0,0 +1,98 @@
+}
+
+// getAttributes creates a lookup of allowed attributes to valid Loki label names.
+func (c *LabelsConfig) getAttributes(labels map[string]string) map[string]model.LabelName {
```
This already existed before the refactoring. Given it will be removed in a few version and this PR is already big as it is, I'd leave it like that.

jpkrohling commented on exporter/lokiexporter/factory_next.go:
```diff
@@ -0,0 +1,39 @@
+func createNextLogsExporter(ctx context.Context, set component.ExporterCreateSettings, cfg *Config) (component.LogsExporter, error) {
+	exp := newNextExporter(cfg, set.TelemetrySettings)
+
+	return exporterhelper.NewLogsExporter(
```
You mean, because of the removal of the attributes? Other than that, we shouldn't be mutating the data. In fact, I would even argue that we shouldn't change the original data at all. I'll change it to reflect the current state, but we likely want to avoid changing the input data.

jpkrohling replied:
I ended up making copies of the resource and log attributes, so that this component is not mutating the incoming data anymore.

bogdandrutu replied:
FYI: That is less efficient. marking mutating will affect performance only if you have multiple pipelines, otherwise the collector is smart enough to not do extra copies.

jpkrohling replied:
Got it, I'll leave it like that and revert if we see performance degradation high enough to warrant mutating the data in the component. I believe that having it semantically correct is preferable to having it with the best performance possible if they contradict each other.

jpkrohling committed:
Use RemoveIf to remove attributes

This also adds the deprecation notes to types.

Signed-off-by: Juraci Paixão Kröhling <juraci@kroehling.de>

bogdandrutu reviewed (approved):
I could not carefully check all the changes since the PR contains a mix of refactoring and new code and hard to identify, but I trust the maintainers of this component that they did their job. 

**Update** I do not "own" this LGTM, this is only to pass the required check, @kovrus @gramidt are the ones who own this :)

jpkrohling commented:
> Update I do not "own" this LGTM, this is only to pass the required check, @kovrus @gramidt are the ones who own this :)

I call this "approval by proxy" :-) Thanks for your comments anyway, they were helpful in making this PR better.
